---
title: 'alignparse: A Python package for parsing complex features out of sequence alignments'
tags:
  - Python
  - PacBio
  - deep mutational scanning
  - single-cell virus sequencing
  - genomics
  - sequencing
authors:
  - name: Katharine H.D. Crawford
    orcid: 0000-0002-6223-4019
    affiliation: "1, 2"
  - name: Jesse D. Bloom
    orcid: 0000-0003-1267-3408
    affiliation: "1, 3"
affiliations:
  - name: Basic Sciences and Computational Biology, Fred Hutchinson Cancer Research Center, Seattle, Washington, USA
    index: 1 
  - name: Department of Genome Sciences and Medical Scientist Training Program, University of Washington, Seattle, Washington, USA
    index: 2
  - name: Howard Hughes Medical Institute, Seattle, Washington, USA
    index: 3
date: 19 November 2019
bibliography: paper.bib
---

# Summary & Purpose

Advances in sequencing technology have made it possible to generate large numbers of long, high-accuracy sequencing reads.
For instance, the new PacBio Sequel platform can generate about 200,000 high-quality circular consensus sequences in a single run [@Hebert:2018].
Numerous programs exist for aligning these reads for purposes of genome assembly.
However, these long reads can also be used for other purposes, such as sequencing long PCR amplicons that contain various sequence features of interest.
For instance, PacBio circular consensus sequences have been used to sequence influenza viruses in single cells [@Russell:2019], or link barcodes to gene mutants in deep mutational scanning [@Matreyek:2018].
For such applications, the alignment of the sequence to the template may be fairly trivial, but it is not trivial to then parse specific features of interest (such mutations, unique molecular identifiers, cell barcodes, and flanking sequences) from these alignments.

Here we describe ``alignparse``, a Python package for parsing complex sets of features from long sequences that map to known amplicons.
This package provides researchers a flexible tool for aligning long-read sequencing data to user-specified target sequences and for extracting short sequences--corresponding to features--from these reads for further analyses. 
Specifically, it allows the user to provide a complex 'target' template in Genbank format that contains an arbitrary number of user-defined sequence features. 
It then aligns the user's long sequencing reads to this target and filters alignments based on whether the user-specified features are present with the desired identity (which is also user-specified). 
It then parses out the sequences, mutations, and/or accuracy of these features as specified by the user. The flexibility built into this package allows it to be used by any researcher looking to parse short sequence features from long sequencing reads that map to pre-defined amplicons. 


# Uses & Examples 

``alignparse`` is designed so that researchers from many areas of genomics can use this tool to parse complex features from data generated by the long-read sequencing of known amplicons.
Below are two known use cases:

## Sequencing deep mutational scanning libraries

In deep mutaitonal scanning experiments, researchers use mutant libraries to assay the effects of thousands of individual mutations to a gene-of-interest in one experiment. 
Often times, any given sequence in a mutant library will contain more than one mutation. 
Long-read sequencing of mutant libraries has allowed researchers conducting deep mutational scanning experiments to link all of the individual mutations in each variant in their library with a unique molecular barcode [@Matreyek:2018]. 
This has facilitated the study of the effects of genetic variants on gene function by improving the data generated by deep mutational scanning experiments. 
The ``alignparse`` package can be used to parse barcodes and their linked mutations from the long-read sequencing of deep mutational scanning libraries. 
The ability to easily define any number of sequence features is very useful for customizing this sequence analysis, and the ``alignparse.consensus`` submodule provides important tools for determining the consensus sequence from a set of reads that share a barcode.

Currently, ``alignparse`` is being used to analyze deep mutational scanning libraries in the Bloom lab. An example analysis of a [RecA deep mutational scanning library](https://jbloomlab.github.io/alignparse/recA_DMS.html) is included in the documentation.

## Single-cell viral sequencing

Given their small size, several viral genomes can be sequenced in full using long-read sequencing technology.
Indeed, for influenza virus, each of its eight gene segments can be sequenced as a single long-read sequencing amplicon.
Using long-read sequencing in conjunction with single cell technologies (such as 10x Chromium), it is possible to barcode single flu virions and sequence the entire genome of a virus that enters a single cell.
To analyze these viral sequences, the viral genomic sequences and associated mutations must be parsed in conjunction with cell-specific barcodes and additional sequence features introduced to ensure accurate variant calling.
Such experiments and analyses have been carried out for influenza virus, improving our understanding of influenza infection heterogeneity [@Russell:2019].
The [Single-cell virus sequencing](https://jbloomlab.github.io/alignparse/flu_virus_seq_example.html) example shows how data from these experiments can be readily analyzed using ``alignparse``. 



# How ``alignparse`` Works

The main functionality of the ``alignparse`` package is to parse features--provided in a user-defined Genbank file--from alignments of long sequencing reads aligned to a user-specified reference. ``alignparse`` uses [minimap2](https://github.com/lh3/minimap2) for aligning sequences and takes advantage of the short [cs tags](https://lh3.github.io/minimap2/minimap2.html#10) generated by such alignments to help parse features [Li:2018]. 

The [alignparse.minimap2](https://jbloomlab.github.io/alignparse/alignparse.minimap2.html) submodule runs [minimap2](https://github.com/lh3/minimap2) and provides alignment settings optimized for the two use cases described above.
The [alignparse.targets](https://jbloomlab.github.io/alignparse/alignparse.targets.html) submodule includes the majority of the functionality of this package. 
It makes an alignment of long-read squencing data and user-defined target sequences, and parses out user-specified features, filtering out reads with features that fail to meet user-defined sequence parameters.
Further analyses of parsed features are facilitated by [alignparse.consensus](https://jbloomlab.github.io/alignparse/alignparse.consensus.html).
This submodule provides tools for grouping reads by shared barcodes, determining consensus sequences for barcoded reads, and further processing mutation information for downstream analyses.
Downstream analyses of reads parsed using ``alignparse`` can be highly customized by the user as the main outputs from this module are in intuitive data frame formats that can be easily used as input for additional analyses. 
Thus, ``alignparse`` provides a flexible and useful tool designed specifically for parsing any number of complex features from long-read sequencing data of pre-defined amplicons.

# Acknowledgements

We would like to thank members of the Bloom lab for helpful discussions and beta testing. We also thank ___ **funding sources?**

# References

